### ISSUE TITLE
Latency performance about LowerTriangularMask and NoneType attention bias

### ISSUE THREAD

**Issue by wimh966** [facebookresearch/xformers#924]
# ❓ Questions and Help
Hello! I'm running the memory efficient attention benchmark and I hold some questions about the attention bias type. Could you offer some help to me?

I compare the between using xformers.ops.LowerTriangularMask and None. And I find that the latency of the cutclass operator can decrease much with the LowerTriangularMask which means causal=True, while the flashattention can not.  For example, see the figure below, where the last two columns indicate the latency and speedup, respectively. 
<img width="915" alt="image" src="https://github.com/facebookresearch/xformers/assets/52482918/f3e6f28f-7a07-49b6-9aac-2966cfe19f1d">
I find the speedup of flashattention goes down quickly with LowerTriangularMask Do you have any insights for this? Does the cutlass operator take some special optimization for such a case? Btw, when I check the figures in FlashAttention-v2 paper, I find things do not go in this way.

Thank you for your help in advance.

**Comment by danthe3rd** [comment]
Hi,
Which version of xformers are you using?
I ran a benchmark with the first case (`1-2048-72-128`) and I get different latency numbers (A100 80G):
```
cutlass: 1343.3 us
flash: 714.9 us
cutlass[causal]: 758.7 us
flash[causal]: 530.6 us
```

Repro code:
```python
import torch
import time
import xformers.ops.fmha as fmha

def benchmark_fn(name, fn, *args, **kwargs):
    g = torch.cuda.CUDAGraph()
    fn(*args, **kwargs)
    with torch.cuda.graph(g):
        fn(*args, **kwargs)
    torch.cuda.synchronize()
    begin = time.time()
    reps = 10
    for _ in range(10):
        g.replay()
    torch.cuda.synchronize()
    dt = (time.time() - begin)
    dt_us = int(dt * 1000000) / reps
    print(f"{name}:", dt_us, "us")

Q, K, V = [
    torch.randn([1, 2048, 72, 128], device="cuda", dtype=torch.float16)
    for _ in range(3)
]
benchmark_fn("cutlass", fmha.memory_efficient_attention, Q, K, V, attn_bias=None, op=(fmha.cutlass.FwOp, None))
benchmark_fn("flash", fmha.memory_efficient_attention, Q, K, V, attn_bias=None, op=(fmha.flash.FwOp, None))
benchmark_fn("cutlass[causal]", fmha.memory_efficient_attention, Q, K, V, attn_bias=fmha.attn_bias.LowerTriangularMask(), op=(fmha.cutlass.FwOp, None))
benchmark_fn("flash[causal]", fmha.memory_efficient_attention, Q, K, V, attn_bias=fmha.attn_bias.LowerTriangularMask(), op=(fmha.flash.FwOp, None))
```

**Comment by wimh966** [comment]
Hi, I'm using 0.0.22.post7 installed by pip3 install -U xformers --index-url https://download.pytorch.org/whl/cu121. 
In fact, I find the result of the first case 1-2048-72-128 looks a bit unusual, but others look normal. Is this normal that sometimes FlashAttention might give a slower result?

**Comment by wimh966** [comment]
Besides, from you results, I find that with causal=True and lowertriangularmask, the performance of flashattention can not be improved much while the cutlass can. I'm curious about this point. Do you have any insight for this?

```
cutlass: 1343.3 us
flash: 714.9 us
cutlass[causal]: 758.7 us
flash[causal]: 530.6 us
```

**Comment by danthe3rd** [comment]
Flash-Attention v2 in my experience is always faster than cutlassF. I re-ran the numbers with my benchmark script (still on A100 80G) and the causal version is always faster - not exactly 2x because there is some fixed overhead, but still much faster than in your benchmarks:
```
1-2048-72-128
  cutlass: 1344.3 us
  flash: 715.4 us
  cutlass[causal]: 766.4 us
  flash[causal]: 534.9 us
1-2048-96-128
  cutlass: 1758.5 us
  flash: 963.2 us
  cutlass[causal]: 993.0 us
  flash[causal]: 679.8 us
1-2048-20-256
  cutlass: 982.7 us
  flash: 453.5 us
  cutlass[causal]: 607.9 us
  flash[causal]: 357.6 us
1-2048-80-64
  cutlass: 952.2 us
  flash: 450.2 us
  cutlass[causal]: 551.4 us
  flash[causal]: 375.2 us
```

**Comment by wimh966** [comment]
Hello. Thank you for reply. But there might be some misunderstanding about the question between us. What I want to ask is the speedup of flashattention with and without causal=True. I find with causal=True, the speed of cutlass can be be improved much like from 1344.3 -> 766.4 (roughly half), whereas the flashattention can not be improved that much 766.4 -> 534.9. Thus, with causal=True, the speedup of Flashattention compared to Cutlass will decrease in most of the cases you provide. Could you provide some insights about this? Because I find in their original paper, things do not go in this way.

**Comment by danthe3rd** [comment]
I don't think the Flash-Attention v2 paper promises a 2x speedup with causal=True. In fact taking this graph from the paper:
<img width="327" alt="image" src="https://github.com/facebookresearch/xformers/assets/43445237/1ae8262e-87b0-4901-929a-e40c500230ea">

At seqlen=2k, we 227 TFlops for `causal=False`, and 187 TFlops for `causal=True`. This means that the speedup should be ~1.65 instead of 2x.

The shapes they tested on their paper were `8-2048-16-128` I think. For these I get a speedup of ~1.4x:
```
8-2048-16-128
  cutlass: 2294.2 us (119 TFlops)
  flash: 1233.2 us (222 TFlops)
  cutlass[causal]: 1294.3 us (106 TFlops)
  flash[causal]: 885.7 us (155 TFlops)
```

I'll try to see where this is coming from ...

**Comment by wimh966** [comment]
Yes, it's not 2x. But I find the cutclass using with repo can really achieve 2x with causal=True. Do you add some special optimization for it? Or it's just cutclass can be more adapted to causal=True.

**Comment by danthe3rd** [comment]
So I ran the code again with the code from the Flash-Attention repo (v2.3.2):
```
8-2048-16-128
  flash: 1217.5 us (225 TFlops)
  flash[causal]: 879.0 us (156 TFlops)
```
According to the [Flash-Attention paper](https://arxiv.org/pdf/2307.08691.pdf) we should get 227 TFlops (which we repro here), and 187 TFlops for the causal case - but with my repro I only get 155 TFlops. @tridao do you have an idea where the gap might come from? (This is on A100 / f16)

Repro:
```python
import torch
import time
from flash_attn import flash_attn_func

def benchmark_fn(flop, name, fn, *args, **kwargs):
    g = torch.cuda.CUDAGraph()
    fn(*args, **kwargs)
    with torch.cuda.graph(g):
        fn(*args, **kwargs)
    torch.cuda.synchronize()
    begin = time.time()
    reps = 10
    for _ in range(10):
        g.replay()
    torch.cuda.synchronize()
    dt = (time.time() - begin)
    dt_us = int(dt * 1000000) / reps
    tflops_measured = int((flop / (dt / reps)) / 1e12)
    print(f"{name}:", dt_us, "us", f"({tflops_measured} TFlops)")

for B, M, H, K in [\
    [8, 2048, 16, 128],
]:
    print(f"{B}-{M}-{H}-{K}")
    flops = B*H*(4*M*M*K)
    Q, K, V = [
        torch.randn([B, M, H, K], device="cuda", dtype=torch.float16)
        for _ in range(3)
    ]
    benchmark_fn(flops, "  flash", flash_attn_func, Q, K, V)
    benchmark_fn(flops // 2, "  flash[causal]", flash_attn_func, Q, K, V, causal=True)
```

**Comment by danthe3rd** [comment]
> Yes, it's not 2x. But I find the cutclass using with repo can really achieve 2x with causal=True. Do you add some special optimization for it? Or it's just cutclass can be more adapted to causal=True.

I don't think there are particular optimizations going on there.

**Comment by wimh966** [comment]
I see! So it relates with the implementation of flashattention here that it can not achieve a similar high speedup with causal=True. Thank you so much for your help. I really appreciate that. By the way, may I kindly ask that which tool are you using to get the memory chart figure? I think it shows the memory transfer, which can be critical for kernel implementation?

**Comment by danthe3rd** [comment]
Yes sure, I'm using `ncu` (NSight Compute) to generate a profile, with a command like:
```
/usr/local/cuda-12.1/bin/ncu --kernel-name regex:flash --force-overwrite \
    -o prof.ncu-rep --import-source yes --target-processes all\
    --section ComputeWorkloadAnalysis --section InstructionStats --section LaunchStats --section MemoryWorkloadAnalysis \
    --section Occupancy --section SchedulerStats --section SourceCounters --section SpeedOfLight_HierarchicalTensorRooflineChart \
    --section WarpStateStats --section SpeedOfLight_RooflineChart --section MemoryWorkloadAnalysis_Chart --section MemoryWorkloadAnalysis_Tables \
    python YOUR_SCRIPT.py
```
(with `regex:flash` parameter, it will only profile kernels from Flash-Attention)

You can then open the trace with NVIDIA Nsight Compute GUI to have a lot of profiling information, including (but not limited to) this memory chart.

**Comment by danthe3rd** [comment]
Btw the performance regression in causal might have been addressed in https://github.com/Dao-AILab/flash-attention/commit/b4bf9cc1f309df4664fe50322e90121238d673da, we will update Flash in xFormers :)

**Comment by tridao** [comment]
Yeah sorry there was a performance regression that's now fixed.

**Comment by wimh966** [comment]
Hi, thank you so much for your efforts.
